src/io/unix/sync.rs

```diff
@@ -0,0 +1,197 @@
+        }
+    }
+
+    pub(crate) async fn write_all_at(&self, offset: u64, buf: Bytes) -> IOResult<()> {
+        let file = self.no_lock_fd.clone();
+        if buf.len() <= MAX_SYNC_OPERATION_SIZE {
+            Self::inplace_sync_call(move || file.write_all_at(&buf, offset))
+        } else {
+            Self::background_sync_call(move || file.write_all_at(&buf, offset)).await
+        }
+    }
+
+    #[cfg(not(feature = "async-io-rio"))]
+    pub(crate) async fn read_all(&self) -> Result<BytesMut> {
+        self.read_exact_at_allocate(self.size().try_into()?, 0)
+            .await
+    }
+
+    #[cfg(not(feature = "async-io-rio"))]
+    pub(crate) async fn read_exact_at_allocate(
+        &self,
+        size: usize,
+        offset: u64,
+    ) -> Result<BytesMut> {
+        let buf = BytesMut::zeroed(size);
+        self.read_exact_at(buf, offset).await
+    }
+
+    pub(crate) async fn read_exact_at(&self, mut buf: BytesMut, offset: u64) -> Result<BytesMut> {
+        let file = self.no_lock_fd.clone();
+
+        Ok(if buf.len() <= MAX_SYNC_OPERATION_SIZE {
+            Self::inplace_sync_call(move || file.read_exact_at(&mut buf, offset).map(|_| buf))
+        } else {
+            Self::background_sync_call(move || file.read_exact_at(&mut buf, offset).map(|_| buf))
+                .await
+        }?)
+    }
+
+    pub(crate) async fn fsyncdata(&self) -> IOResult<()> {
+        let fd = self.no_lock_fd.clone();
+        Self::background_sync_call(move || fd.sync_all()).await
+    }
+
+    pub(crate) fn created_at(&self) -> Result<SystemTime> {
+        let metadata = self.no_lock_fd.metadata()?;
+        Ok(metadata.created().unwrap_or(SystemTime::now()))
+    }
+
+    fn advisory_write_lock_file(fd: i32) -> LockAcquisitionResult {
+        let flock = libc::flock {
+            l_len: 0, // 0 means "whole file"
+            l_start: 0,
+            l_whence: libc::SEEK_SET as i16,
+            l_type: libc::F_WRLCK as i16,
+            l_pid: -1, // pid of current file owner, if any (when fcntl is invoked with F_GETLK)
+        };
+        let res = nix::fcntl::fcntl(fd, FcntlArg::F_SETLK(&flock));
+        if let Err(e) = res {
+            warn!("acquiring writelock failed, errno: {:?}", e);
+            match e {
+                Errno::EACCES | Errno::EAGAIN => LockAcquisitionResult::AlreadyLocked,
+                e => LockAcquisitionResult::Error(e),
+            }
+        } else {
+            LockAcquisitionResult::Acquired
+        }
+    }
+
+    async fn background_sync_call<F, R>(f: F) -> R
+    where
+        F: FnOnce() -> R + Send + 'static,
+        R: Send + 'static,
+    {
+        tokio::task::spawn_blocking(move || f())
+            .await
+            .expect("spawned blocking task failed")
+    }
+
+    fn inplace_sync_call<F, R>(f: F) -> R
+    where
+        F: FnOnce() -> R + Send + 'static,
+        R: Send + 'static,
+    {
+        match tokio::runtime::Handle::current().runtime_flavor() {
```

Comment:
This is a bad approach. It is better to avoid `inplace_sync_call` at the caller level when runtime is `CurrentThread`. I suggest to add function `can_run_io_in_place(io_len: usize) -> bool` that checks both  `io_len <= MAX_SYNC_OPERATION_SIZE` and `runtime_flavor() != RuntimeFlavor::CurrentThread`